src/Libraries/RevitNodes/Application/Document.cs

```diff
@@ -170,58 +147,58 @@ private List<ElementId> PurgeMaterials()
+                appearanceAssetIds.Remove(material.AppearanceAssetId);
             }
 
-            if (appearanceAssets.Count > 0)
-                this.InternalDocument.Delete(appearanceAssets);
+            if (appearanceAssetIds.Count > 0)
+                this.InternalDocument.Delete(appearanceAssetIds);
             if (propertySet.Count > 0)
                 this.InternalDocument.Delete(propertySet);
 
-            List<ElementId> purgedElementIds = new List<ElementId>();
-            purgedElementIds.AddRange(appearanceAssets);
-            purgedElementIds.AddRange(propertySet);
+            List<ElementId> purgedElementIds = new List<ElementId>(appearanceAssetIds.Concat(propertySet));
             return purgedElementIds;
         }
 
-        private List<ElementId> PurgeElements(List<PerformanceAdviserRuleId> performanceAdviserRuleId, bool runRecursively)
+        private List<ElementId> PurgeElements(bool runRecursively)
         {
             List<ElementId> purgedElementIds = new List<ElementId>();
-            List<ElementId> purgableElementIds = GetPurgeableElementIds(performanceAdviserRuleId);
+            List<ElementId> purgeableElementIds = GetPurgeableElementIds();
             
             if (!runRecursively)
             {
-                if (purgableElementIds == null)
+                if (purgeableElementIds == null)
                     return purgedElementIds;
-                purgedElementIds.AddRange(purgableElementIds);
-                this.InternalDocument.Delete(purgableElementIds);
+                purgedElementIds.AddRange(purgeableElementIds);
+                this.InternalDocument.Delete(purgeableElementIds);
                 return purgedElementIds;
             }
-            if (purgableElementIds == null)
+            if (purgeableElementIds == null)
                 return purgedElementIds;
 
-            purgedElementIds.AddRange(purgableElementIds);
-            this.InternalDocument.Delete(purgableElementIds);
-            purgedElementIds.AddRange(PurgeElements(performanceAdviserRuleId, runRecursively));
+            purgedElementIds.AddRange(purgeableElementIds);
+            this.InternalDocument.Delete(purgeableElementIds);
+            purgedElementIds.AddRange(PurgeElements(runRecursively));
 
             return purgedElementIds;
         }
 
-        private List<Autodesk.Revit.DB.ElementId> GetPurgeableElementIds(List<PerformanceAdviserRuleId> performanceAdviserRuleId)
+        private List<Autodesk.Revit.DB.ElementId> GetPurgeableElementIds()
         {
             List<Autodesk.Revit.DB.FailureMessage> failureMessages = PerformanceAdviser.GetPerformanceAdviser()
-                                                                                       .ExecuteRules(this.InternalDocument, performanceAdviserRuleId)
+                                                                                       .ExecuteRules(this.InternalDocument, GetPerfromanceAdviserRuleId())
```

Comment:
we're calling `GetPerfromanceAdviserRuleId` multiple types (recursively) and in theory it should return the same thing each time, we should cache this in a private property on the class and only interact with Revit DB if needed